pkg/operator/apiserver/controller/workload/workload.go

```diff
@@ -0,0 +1,457 @@
+	versionRecorder status.VersionGetter) *Controller {
+	controllerRef := &Controller{
+		operatorNamespace:            operatorNamespace,
+		name:                         fmt.Sprintf("%sWorkloadController", name),
+		targetNamespace:              targetNamespace,
+		targetOperandVersion:         targetOperandVersion,
+		operandNamePrefix:            operandNamePrefix,
+		operatorClient:               operatorClient,
+		kubeClient:                   kubeClient,
+		delegate:                     delegate,
+		openshiftClusterConfigClient: openshiftClusterConfigClient,
+		eventRecorder:                eventRecorder.WithComponentSuffix("workload-controller"),
+		versionRecorder:              versionRecorder,
+		queue:                        workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), name),
+	}
+
+	return controllerRef
+}
+
+func (c *Controller) sync() error {
+	operatorSpec, _, _, err := c.operatorClient.GetOperatorState()
+	if err != nil {
+		return err
+	}
+
+	if run, err := c.shouldSync(operatorSpec); !run {
+		return err
+	}
+
+	if fulfilled, err := c.preconditionFulfilled(operatorSpec); !fulfilled {
+		return err
+	}
+
+	if fulfilled, err := c.delegate.PreconditionFulfilled(); !fulfilled {
+		return err
+	}
+
+	workload, operatorConfigAtHighestGeneration, errs := c.delegate.Sync()
+
+	return c.updateOperatorStatus(workload, operatorConfigAtHighestGeneration, errs)
+}
+
+// Run starts workload controller and blocks until stopCh is closed.
+// Note that setting workers doesn't have any effect, the controller is single-threaded.
+func (c *Controller) Run(ctx context.Context, workers int) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	klog.Infof("Starting %s", c.name)
+	defer klog.Infof("Shutting down %s", c.name)
+	if !cache.WaitForCacheSync(ctx.Done(), c.preRunCachesSynced...) {
+		utilruntime.HandleError(fmt.Errorf("caches did not sync"))
+		return
+	}
+
+	// doesn't matter what workers say, only start one.
+	go wait.Until(c.runWorker, time.Second, ctx.Done())
+
+	<-ctx.Done()
+}
+
+func (c *Controller) runWorker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *Controller) processNextWorkItem() bool {
+	dsKey, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(dsKey)
+
+	err := c.sync()
+	if err == nil {
+		c.queue.Forget(dsKey)
+		return true
+	}
+
+	utilruntime.HandleError(fmt.Errorf("%v failed with : %v", dsKey, err))
+	c.queue.AddRateLimited(dsKey)
+
+	return true
+}
+
+// AddInformer queues the given informer to check spec, status and managed resources
+func (c *Controller) AddInformer(informer cache.SharedIndexInformer) *Controller {
+	informer.AddEventHandler(c.eventHandler())
+	c.preRunCachesSynced = append(c.preRunCachesSynced, informer.HasSynced)
+	return c
+}
+
+// AddNamespaceInformer queues the given ns informer for the targetNamespace
+func (c *Controller) AddNamespaceInformer(informer cache.SharedIndexInformer) *Controller {
```

Comment:
how odd that we have a special namespace informer code path. @mfojtik do you have that on the radar?